applications/CompressiblePotentialFlowApplication/custom_elements/embedded_incompressible_potential_flow_element.cpp

```diff
@@ -0,0 +1,178 @@
+// Public Operations
+
+template <int Dim, int NumNodes>
+Element::Pointer EmbeddedIncompressiblePotentialFlowElement<Dim, NumNodes>::Create(
+    IndexType NewId, NodesArrayType const& ThisNodes, typename PropertiesType::Pointer pProperties) const
+{
+    KRATOS_TRY
+    return Kratos::make_shared<EmbeddedIncompressiblePotentialFlowElement>(
+        NewId, this->GetGeometry().Create(ThisNodes), pProperties);
+    KRATOS_CATCH("");
+}
+
+template <int Dim, int NumNodes>
+Element::Pointer EmbeddedIncompressiblePotentialFlowElement<Dim, NumNodes>::Create(
+    IndexType NewId, typename GeometryType::Pointer pGeom, typename PropertiesType::Pointer pProperties) const
+{
+    KRATOS_TRY
+    return Kratos::make_shared<EmbeddedIncompressiblePotentialFlowElement>(
+        NewId, pGeom, pProperties);
+    KRATOS_CATCH("");
+}
+
+template <int Dim, int NumNodes>
+Element::Pointer EmbeddedIncompressiblePotentialFlowElement<Dim, NumNodes>::Clone(
+    IndexType NewId, NodesArrayType const& ThisNodes) const
+{
+    KRATOS_TRY
+    return Kratos::make_shared<EmbeddedIncompressiblePotentialFlowElement>(
+        NewId, this->GetGeometry().Create(ThisNodes), this->pGetProperties());
+    KRATOS_CATCH("");
+}
+
+template <int Dim, int NumNodes>
+void EmbeddedIncompressiblePotentialFlowElement<Dim, NumNodes>::CalculateLocalSystem(
+    MatrixType& rLeftHandSideMatrix, VectorType& rRightHandSideVector, ProcessInfo& rCurrentProcessInfo)
+{
+    const EmbeddedIncompressiblePotentialFlowElement& r_this = *this;
+    const int wake = r_this.GetValue(WAKE);
+    const int kutta = r_this.GetValue(KUTTA);
+
+
+    if (this->Is(BOUNDARY) && wake == 0 && kutta == 0)
+        CalculateEmbeddedLocalSystem(rLeftHandSideMatrix,rRightHandSideVector,rCurrentProcessInfo);
+    else
+        BaseType::CalculateLocalSystem(rLeftHandSideMatrix, rRightHandSideVector, rCurrentProcessInfo);
+
+}
+
+template <int Dim, int NumNodes>
+void EmbeddedIncompressiblePotentialFlowElement<Dim, NumNodes>::CalculateEmbeddedLocalSystem(
+    MatrixType& rLeftHandSideMatrix, VectorType& rRightHandSideVector, ProcessInfo& rCurrentProcessInfo)
+{
+    if (rLeftHandSideMatrix.size1() != NumNodes || rLeftHandSideMatrix.size2() != NumNodes)
+        rLeftHandSideMatrix.resize(NumNodes, NumNodes, false);
+    if (rRightHandSideVector.size() != NumNodes)
+        rRightHandSideVector.resize(NumNodes, false);
+    rLeftHandSideMatrix.clear();
+
+    array_1d<double, NumNodes> potential;
+    array_1d<double, NumNodes> elemental_distance;
+    for(unsigned int i_node = 0; i_node<NumNodes; i_node++)
+        elemental_distance[i_node] = this->GetGeometry()[i_node].GetSolutionStepValue(LEVEL_SET);
+
+    BaseType::GetPotentialOnNormalElement(potential);
+
+    const Vector& r_elemental_distances=elemental_distance;
+    Triangle2D3ModifiedShapeFunctions triangle_shape_functions(this->pGetGeometry(), r_elemental_distances);
```

Comment:
I'd create a method that returns a pointer to the modified shape functions. This method now will be a specialization for the triangle but in the future, as soon as you move to the tetrahedron, it will have another specialization for calling the `Tetrahedra3D4ModifiedShapeFunctions`.

Reply:
Ok! I added a function called pGetModifiedShapeFunctions, that gives an error if Dim is not equal to 2.

Reply:
I've done a suggested change. Have a look to it. The idea is to take advantage of the template implementation to specialize the method to avoid the if statements.